presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewCandidateExtractor.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.common.QualifiedObjectName;
+import com.facebook.presto.sql.tree.DefaultTraversalVisitor;
+import com.facebook.presto.sql.tree.Table;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.metadata.MetadataUtil.createQualifiedObjectName;
+import static java.util.Objects.requireNonNull;
+
+public class MaterializedViewCandidateExtractor
+        extends DefaultTraversalVisitor<Void, Void>
+{
+    private final Set<QualifiedObjectName> tableNames = new HashSet<>();
+    private final Map<QualifiedObjectName, List<QualifiedObjectName>> baseTableToMaterializedViews;
+
+    private final Session session;
+
+    public MaterializedViewCandidateExtractor(Session session, Map<QualifiedObjectName, List<QualifiedObjectName>> baseTableToMaterializedViews)
+    {
+        this.session = requireNonNull(session, "session is null");
+        this.baseTableToMaterializedViews = requireNonNull(baseTableToMaterializedViews, "baseTableToMaterializedViews is null");
+    }
+
+    @Override
+    protected Void visitTable(Table node, Void context)
+    {
+        tableNames.add(createQualifiedObjectName(session, node, node.getName()));
+        return null;
+    }
+
+    public Optional<QualifiedObjectName> getMaterializedViewCandidate()
+    {
+        Set<QualifiedObjectName> materializedViewCandidates = new HashSet();
+
+        for (QualifiedObjectName baseTable : tableNames) {
+            List<QualifiedObjectName> materializedViews = baseTableToMaterializedViews.getOrDefault(baseTable, new ArrayList<>());
+
+            if (materializedViewCandidates.isEmpty()) {
+                materializedViewCandidates.addAll(materializedViews);
+            }
+            else {
+                materializedViewCandidates.retainAll(materializedViews);
+            }
+
+            if (materializedViewCandidates.isEmpty()) {
+                return Optional.empty();
+            }
+        }
+
+        // Temporarily it returns the first materialized view
+        // TODO: Find the appropriate materialized view for the base query
+        return Optional.of(materializedViewCandidates.iterator().next());
```

Comment:
This looks suboptimal. We should return all candidates materialized views for all tables in the current query.

Reply:
Are we sure CandidateExtractor will have the full context of what a proper materialized view would be? To achieve that, it is pretty complicated. Is the goal to have some pattern matching within CandidateExtractor?

Reply:
@highker not sure if I follow your concern. 
Would figuring out the best candidate logic be complicated? yes it would be.
But the only information we need to find a candidate is following
1. Original query
2. List of MVs supported by the base tables used in the query

The current candidate extractor have both information, so it should be able to do the necessary work(create more processor) to find the candidate.

If you feel strongly about it, we can return the list of candidates for now, and change the interface later.

Reply:
Yes, let's change it to a list to make this commit more consistent. Otherwise, this PR LGTM